### PR TITLE
【2.0 API】Add CUDA kernel and enhance options for grid_sample 

### PR DESCRIPTION
**Pull request by wanghaoshuang**
### PR types

New features

### PR changes

OPS

### Describe

This PR enhance CPU kernel and add new CUDA kernel to make grid_sample support:
1. align_corners: with bool type.
2. padding mode: which can be in ['zeros', 'reflect', 'border']
3. Interpolation mode:  which ca be in ['bilinear', 'nearest']

The old CPU and CUDNN version only support align_corners=true, padding_mode='zeros' and interpolation_mode='bilinear'.

The behavior of the new version op in default mode is compatible with the old version.

### CODE REVIEW

File: paddle/fluid/operators/grid_sampler_op.cu [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+#include <algorithm>
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/grid_sampler_op.h"
+#include "paddle/fluid/platform/cuda_device_function.h"
+#include "paddle/fluid/platform/gpu_info.h"
+
+namespace paddle {
+namespace operators {
+
+using Mode = paddle::operators::Model;
```
Comment (willthefrog) [inline paddle/fluid/operators/grid_sampler_op.cu]: `Model`?

File: python/paddle/nn/functional/vision.py [hunk 2]
```diff
@@ -89,3 +88,157 @@
...
+
+    Returns:
+        Tensor: Shape of output is [N, C, grid_H, grid_W] in which `grid_H` is the height of grid
+                and `grid_W` is the width of grid. The data type is same as input tensor.
+
+    Examples:
+
+        .. code-block:: python
+
+            import paddle.fluid as fluid
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: update sample

File: paddle/fluid/operators/grid_sampler_op.h [hunk 3]
```diff
@@ -22,6 +25,13 @@ limitations under the License. */
 namespace paddle {
 namespace operators {
 
+enum class Model {
```
Comment (willthefrog) [inline paddle/fluid/operators/grid_sampler_op.h]: `Model`?

File: python/paddle/nn/functional/vision.py [hunk 4]
```diff
@@ -89,3 +88,157 @@
...
+
+            # use with affine_grid
+            x = fluid.data(name='x', shape=[None, 10, 32, 32], dtype='float32')
+            theta = fluid.layers.data(name='theta', shape=[2, 3], dtype='float32')
+            grid = fluid.layers.affine_grid(theta=theta, out_shape=[3, 10, 32, 32])
+            out = fluid.layers.grid_sampler(x=x, grid=grid)
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: `sampler` -> `sample`

File: python/paddle/nn/functional/vision.py [hunk 5]
```diff
@@ -89,3 +88,157 @@
...
+
+    """
+    helper = LayerHelper("grid_sampler", **locals())
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: `sampler` -> `sample`

**Review by heavengate — DISMISSED** [review]

File: paddle/fluid/operators/grid_sampler_op.cc [hunk 6]
```diff
@@ -58,21 +59,10 @@ class GridSampleOp : public framework::OperatorWithKernel {
...
-      PADDLE_ENFORCE_EQ(
-          grid_dims[2], x_dims[3],
-          platform::errors::InvalidArgument(
-              "Input(X) dims[3] and Input(Grid) dims[2] should be equal, but "
-              "received X dimension[3](%d) != Grid dimension[2](%d)",
-              x_dims[3], grid_dims[2]));
     }
 
-    ctx->SetOutputDim("Output", x_dims);
+    ctx->SetOutputDim("Output",
```
Comment (willthefrog) [inline paddle/fluid/operators/grid_sampler_op.cc:64]: make sure not to break backward compatibility, otherwise a new OP should be  added

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/grid_sampler_op.cu (hunk 1 above)
Comment (wanghaoshuang) [reply paddle/fluid/operators/grid_sampler_op.cu]: Fixed.

File: python/paddle/nn/functional/vision.py (hunk 2 above)
Comment (wanghaoshuang) [reply python/paddle/nn/functional/vision.py]: Done.

File: python/paddle/nn/functional/vision.py (hunk 5 above)
Comment (wanghaoshuang) [reply python/paddle/nn/functional/vision.py]: Updated.

File: paddle/fluid/operators/grid_sampler_op.cc (hunk 6 above)
Comment (wanghaoshuang) [reply paddle/fluid/operators/grid_sampler_op.cc:64]: Yes. It is compatibility with old version.

**Review by heavengate — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Comment by raindrops2sea** [comment]
> ### PR types
> New features
> 
> ### PR changes
> OPS
> 
> ### Describe
> Add CUDA kernel and enhance options for grid_sample

a more detailed description is needed.

**Review by willthefrog — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]
LGTM

**Pull request merged**